Cupy support to lambdify

thomasaarholt opened the pull request:
#### References to other Issues or PRs
Resubmission of #20622 (which itself was a resubmission of #20271). 
This PR takes those lessons into account.

#### Brief description of what is fixed or changed

- Refactored the numpy array code to allow other libraries to subclass it
- Moved the numpy/scipy (and cupy) code out of pycode.py and into numpy.py
- Added support for CuPy (very few lines needed)

#### Other comments
CuPy is technically already implicitly supported through numpy's `__array_function__` interface. If you use a lambdify function on a cupy array, it will work without this PR. I think explicit support is still useful, for testing purposes, advertising purposes ("sympy supports cupy!") and because it costs very little to support.

I have assumed that all sympy-relevant numpy functions are supported in cupy. That means that if one isn't a cupy error will be raised. I think this is fine, as A) [most functions are supported already](https://docs.cupy.dev/en/stable/reference/comparison.html) and B) we don't want to maintain this.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* utilities
  * Added official support for using CuPy to GPU accelerate lambdify functions.
<!-- END RELEASE NOTES -->

thomasaarholt committed:
Refactor numpy code, allow other subclassing for other libraries

remove whitespace

numpy white

thomasaarholt committed:
add support for CuPy

cupy white

thomasaarholt committed:
add tests

more tests

test whitespace

newline

remove import

thomasaarholt committed:
add docs

asmeurer commented on sympy/printing/numpy.py:
```diff
@@ -0,0 +1,494 @@
+    logical operators, etc.
+    """
+
+    _module = 'NumPy'
```
Shouldn't this be lowercase?

thomasaarholt replied:
I wanted to preserve the formatting of "NumPy" for the `self.language = "Python with {}".format(self._module)` line further down. You may notice that I call `.lower()` on `_module` a few times, and then overwrite it to be lowercase. 
The following lines, which are called _after_ the `class NumPyPrinter...`, made writing the code rather tricky. If they can be called via the `def __init__()` instead, then I could clean up some parts.
```
# line 315
for k in NumPyPrinter._kf:
    setattr(NumPyPrinter, '_print_%s' % k, _print_known_func)

for k in NumPyPrinter._kc:
    setattr(NumPyPrinter, '_print_%s' % k, _print_known_const)
```

asmeurer replied:
I find that to be very confusing. Where is `self.language` used? I would just make numpy lowercase everywhere.

thomasaarholt replied:
I took the approach of not wanting to change too much unnecessarily out of fear of breaking / stepping on toes. In particular, there were several tests failing with NumPy -> numpy.
I've now edited those tests to be a bit simpler, and gone lowercase everywhere.

asmeurer commented:
It would be good to have more extensive tests. I have a machine that can run cupy that I can run the tests on.

thomasaarholt commented:
Using the SciPyPrinter as an example, I don't fully understand how the following lines work. `_print_known_func` doesn't appear to be given the expr argument that it expects. This chunk of code made it a little bit trickier to subclass the NumPyPrinter.

```python
for k in SciPyPrinter._kf:
    setattr(SciPyPrinter, '_print_%s' % k, _print_known_func)

for k in SciPyPrinter._kc:
    setattr(SciPyPrinter, '_print_%s' % k, _print_known_const)
```

thomasaarholt commented:
Thanks for the quick look, btw :)

thomasaarholt committed:
refactor numpy/scipy/cupy

thomasaarholt committed:
more tests

thomasaarholt commented:
With a fresh evening look, I've refactored a bit of the code that was giving me trouble previously. I think it's now a lot cleaner. I've added some more tests, happy to add more. Ideas? I don't think we need to test that *every* cupy function is formatted correctly, since the cupy wording is identical to the numpy one.

bjodah commented:
I think it looks good. I don't have cupy at hand so I haven't tested this.
If the setattr trick makes subclassing hard perhaps a class decorator could be used instead (I haven't tried though).

thomasaarholt commented:
I'm happy to leave it as it is now. The setattr trick wasn't a problem in the end, after all.
@bjodah will you approve the review?

bjodah commented:
Great. Yes, thank you for this!

asmeurer commented on doc/src/modules/numeric-computation.rst:
```diff
@@ -73,6 +73,18 @@ If you have array-based data this can confer a considerable speedup, on the
+increasing support for other GPUs manufacturer. It can in many cases be used as 
+a drop-in replacement for numpy. It is particu
+
+    >>> f = lambdify(x, expr, "cupy")
```
You need to update the blacklisting of this file in sympy/utilities/runtests.py so that the doctests for it aren't run if cupy isn't installed.

thomasaarholt replied:
I'm happy to do this, but I think I need some help. Based on the contents of /runtests.py, I have no idea how to proceed with blacklisting the rst file.

thomasaarholt replied:
Ah, you meant sympy/testing/runtests.py, not sympy/utilities/runtests.py.

asmeurer replied:
Oh sorry. The file used to live in utilities for a long time, and old habits die hard.

asmeurer commented:
I ran the tests on my machine that has cupy. Everything passes. There is just the one issue with the doctest not being skipped if cupy isn't installed. I guess it currently isn't run on CI (maybe theano isn't installed)? That's actually not great that it doesn't run at all, and the doctester doesn't even indicate that it is skipped. But that is something that is also a problem for other files as well.

thomasaarholt commented:
It's a bit of a shame that blacklisting numeric-computation.rst when cupy is not installed, will mean that CI will never test the file (unless it gets a GPU+cupy).
I note that [sphinx has inline codes](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html#skipping-tests-conditionally) for rst docs that automatically will run (and skip, if condition True) doctests. Sympy does not use these? Doessympy choose to run *all* bits of code in the docs?

thomasaarholt committed:
blacklist numeric-computation.rst when cupy not installed

thomasaarholt commented:
I think this is good to go.

thomasaarholt commented:
I have taken all comments into account.

asmeurer commented:
> It's a bit of a shame that blacklisting numeric-computation.rst when cupy is not installed, will mean that CI will never test the file (unless it gets a GPU+cupy).
I note that sphinx has inline codes for rst docs that automatically will run (and skip, if condition True) doctests. Sympy does not use these? Does sympy choose to test all bits of code in the docs?

I agree. I didn't know about these Sphinx directives. We should look into using those. I seems like that is based on running the doctests through the Sphinx doctest runner (?), which we don't use. But regardless, we should use some sort of directive in the rst files to skip individual doctests so that we don't have to skip entire files. There's also the issue I mentioned that the SymPy doctest runner doesn't even indicate that the file is skipped. So there is some improvement that can be done there. But we should open a separate issue about these things. For now, we should either skip the entire file if cupy isn't installed, or `# doctest: +SKIP` the doctest in question.

asmeurer commented:
Manually confirmed the tests pass with cupy installed.

asmeurer commented:
I opened https://github.com/sympy/sympy/issues/21217 for the doctest issue.
